File: src/0091.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if not s or s[0] == '0':
            return 0
        dp = [0 for _ in range(len(s) + 1)]
        dp[0:2] = [1,1]
        for i in range(2, len(s) + 1):
            if 0 < int(s[i - 1]):
                dp[i] = dp[i - 1]
            if 10 <= int(s[i - 2:i]) <= 26:
                dp[i] += dp[i - 2]
            if dp[i] == dp[i - 1] == 0:
                break
        return dp[-1]

File: src/0091.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int numDecodings(string s) {
        if (s.empty() || s[0] == '0') {
            return 0;
        }

        vector<int> dp (s.size() + 1, 0);
        dp[0] = 1;
        dp[1] = 1;

        for (int i = 2; i < s.size() + 1; ++i) {
            if (0 < s[i - 1] - '0') {
                dp[i] = dp[i - 1];
            }
            int num = (s[i - 2] - '0') * 10 + (s[i - 1] - '0');
            if (10 <= num && num <= 26) {
                dp[i] += dp[i - 2];
            }
            if (dp[i] == 0 && dp[i - 1] == 0) {
                break;
            }
        }
        return dp[s.size()];
    }
};
